### PR TITLE
moves audit event copy to handler

### PR DESCRIPTION
**Pull request by pbarker**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
As a part of the dynamic audit work kubernetes/enhancements#600, we are moving the event copy from the buffer to the handler. This will allow us to have multiple cheap buffers by sharing the event pointer.

**Does this PR introduce a user-facing change?**:
```
NONE
```

### CODE REVIEW

**Comment by roycaihw** [comment]
cc @yliaog

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go [hunk 1]
```diff
@@ -149,8 +149,11 @@ func processAuditEvent(sink audit.Sink, ev *auditinternal.Event, omitStages []au
 	} else {
 		ev.StageTimestamp = metav1.NewMicroTime(time.Now())
 	}
+	// Make deep copy of event before sending to the sink, so that as it can be processed concurrently as
+	// the request passes through its stages
+	e := ev.DeepCopy()
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: Update the godoc on the ProcessEvents() interface to indicate the event may not be mutated, but is safe to use from spawned goroutines

This change means that when multiple sinks exist (logfile+webhook, logfile+dynamic, etc), the same instance of the audit event will be sent to all of their ProcessEvents handlers. This means we need to make sure that no mutation of the original object is done in *any* ProcessEvents methods. 

TruncateBackend modifies the annotations map in the passed audit event (shallow copy keeps the same annotations map):

https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/staging/src/k8s.io/apiserver/plugin/pkg/audit/truncate/truncate.go#L121-L127

Another concerning implication is what happens when the events are converted to external versions and encoded (along with the nested request/response objects). Do we exercise any of the encoding paths that briefly set group/version/kind?
* https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go#L209-L212
* https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go#L262-L264
* https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go#L231-L233

I see us encoding (or making calls that lead to encoding) in these places:
* https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/staging/src/k8s.io/apiserver/plugin/pkg/audit/log/backend.go#L74
* https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/staging/src/k8s.io/apiserver/plugin/pkg/audit/truncate/truncate.go#L141
* https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go#L96

We need to verify and add tests that exercise those backend paths concurrently to make sure we are not introducing data races with this.

**Comment by liggitt** [comment]
agreed, moving to 1.14 to get more thorough tests in place around this change

/milestone v1.14

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go (hunk 1 above)
Comment (pbarker) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: Ok @liggitt, circling back on this. As far as the truncate piece goes, the functionality there hasn’t changed from what was existing. That being said it isn’t desirable behavior. I think the simplest solve may be to copy the annotations in the truncate plugin, I've added this to the PR.

For the encoding piece, the shallow copy seems to take care of that, but I’ve added tests to check that we aren’t racing. Let me know what you think

File: staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic_test.go [hunk 2]
```diff
@@ -176,6 +177,24 @@ func TestDynamic(t *testing.T) {
 		require.NoError(t, err, "unable to find events sent to sink 1")
 		err = checkForEvent(eventList2, testEvent)
 		require.NoError(t, err, "unable to find events sent to sink 2")
+
+		// test that backends don't race
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic_test.go]: Is the idea that the race detector should be triggered if there is a race here? Is there any additional verification that can be done?

Comment (pbarker) [reply staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic_test.go]: yep thats the idea. I added some checks that test that we aren't missing events and that the events received are all consistent. Let me know what you think

**Comment by nikopen** [comment]
This seems to be ready for merge, @liggitt  ?

File: staging/src/k8s.io/apiserver/pkg/audit/types.go [hunk 3]
```diff
@@ -25,7 +25,7 @@ type Sink interface {
 	// Errors might be logged by the sink itself. If an error should be fatal, leading to an internal
 	// error, ProcessEvents is supposed to panic. The event must not be mutated and is reused by the caller
 	// after the call returns, i.e. the sink has to make a deepcopy to keep a copy around if necessary.
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/audit/types.go]: Update this comment:

```diff
- The event must not be mutated and is reused by the caller after the call returns, i.e. the sink has to make a deepcopy to keep a copy around if necessary.
+ The event must not be mutated, i.e. the sink has to make a deepcopy prior to any modifications.
```

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: when I commented out the shallow copy and annotation copy, the async race tests don't fail. are we exercising the backends we actually use in that test?

**Comment by liggitt** [comment]
/hold

until we can prove the async test is actually exercising potential race condition paths noted in https://github.com/kubernetes/kubernetes/pull/70560/files#r263058036

**Comment by soggiest** [comment]
Hello, can we move this to v1.15 similar to #70816 ?

**Comment by pbarker** [comment]
yes siree @soggiest 🙏

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go (hunk 1 above)
Comment (pbarker) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: I'll add another test at the handler, @liggitt would it make sense to feature gate this?

File: staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered.go [hunk 4]
```diff
@@ -271,12 +278,12 @@ func (b *bufferedBackend) ProcessEvents(ev ...*auditinternal.Event) bool {
 
 	for i, e := range ev {
 		evIndex = i
-		// Per the audit.Backend interface these events are reused after being
-		// sent to the Sink. Deep copy and send the copy to the queue.
-		event := e.DeepCopy()
```
Comment (cwdsuzhou) [inline staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered.go]: We may need add the comments back

File: staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic_test.go [hunk 5]
```diff
@@ -317,3 +401,12 @@ func defaultTestConfig() (*Config, chan struct{}) {
...
 	}, stop
 }
+
+// asyncProcess sends the given event the given number of iterations to the backend and is meant to be
+// run concurrently
+func asyncProcess(wg *sync.WaitGroup, numIterations int, event auditinternal.Event, b audit.Backend) {
+	for i := 0; i < numIterations; i++ {
+		b.ProcessEvents(&event)
+	}
+	wg.Done()
```
Comment (cwdsuzhou) [inline staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic_test.go]: I think move the wg.Done() out would improve the readability

**Comment by soggiest** [comment]
Hello! Code Freeze is just about on us. We'll be entering into Freeze tomorrow, May 31st. Is this still planned to land in 1.15?

**Comment by pbarker** [comment]
going to push this one another release @soggiest, thanks

**Comment by ttousai** [comment]
Hello! I'm a bug triage shadow for the 1.16 release cycle and considering this issue is tagged for 1.16, I'd like to check its status. Code freeze is starting on August 29th (about 3 days from now), which means that this PR must merged by then.

Do we still target this issue to be fixed for 1.16?

**Comment by liggitt** [comment]
Closing due to inactivity. Feel free to reopen when ready to resume.

/close

**Pull request closed without merging**